File: week1_programming_challenges/2_maximum_pairwise_product/max_pairwise_product.py
# python3


def max_pairwise_product(numbers):
    n = len(numbers)
    max_product = 0
    for first in range(n):
        for second in range(first + 1, n):
            max_product = max(max_product,
                              numbers[first] * numbers[second])

    return max_product


def max_pairwise_product_fast(numbers):
    first = 0
    second = 0
    for number in numbers:
        if number > first:
            first, second = number, first
        elif second < number <= first:
            second = number
    return first * second


def test_result():
    assert max_pairwise_product([1, 2, 3]) == 6
    assert max_pairwise_product([9, 1, 9]) == 81
    assert max_pairwise_product([100000, 90000]) == 9000000000


if __name__ == '__main__':
    test_result()
    input_n = int(input())
    input_numbers = [int(x) for x in input().split()]
    print(max_pairwise_product_fast(input_numbers))

File: week1_programming_challenges/2_maximum_pairwise_product/max_pairwise_product.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <cassert>

int MaxPairwiseProduct(const std::vector<int>& numbers) {
    int max_product = 0;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = first + 1; second < n; ++second) {
            max_product = std::max(max_product,
                numbers[first] * numbers[second]);
        }
    }

    return max_product;
}

uint64_t MaxPairwiseProductFast(const std::vector<int>& numbers) {
    uint64_t first = 0;
    uint64_t second = 0;

    for (const auto& i : numbers) {
        if (i > first) {
            second = first;
            first = i;
        } else if (i > second && i <=first) {
            second = i;
        }
    }

    return first * second;
}

void TestResult() {
    assert(MaxPairwiseProductFast({1, 2, 3}) == 6);
    assert(MaxPairwiseProductFast({9, 1, 9}) == 81);
    assert(MaxPairwiseProductFast({100000, 90000}) == 9000000000);
}

int main() {
    TestResult();
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProductFast(numbers) << "\n";
    return 0;
}
